app/src/main/java/ai/elimu/appstore/synchronization/AppListAdapter.java

```diff
@@ -190,13 +201,16 @@ public void onBindViewHolder(final ViewHolder holder, final int position) {
                 } catch (PackageManager.NameNotFoundException e) {
                     Timber.e(e, null);
                 }
-            } else if (existingApkFile.exists()) {  //Extract app icon from downloaded APK if found
+            } else if (isExistingAndValidApk(existingApkFile, applicationVersion.getFileSizeInKb())) {
+                //Extract app icon from downloaded APK if found
                 PackageInfo packageInfo = packageManager.getPackageArchiveInfo(existingApkFile
                         .getAbsolutePath(), 0);
-                packageInfo.applicationInfo.sourceDir = existingApkFile.getAbsolutePath();
-                packageInfo.applicationInfo.publicSourceDir = existingApkFile.getAbsolutePath();
-                Drawable appIcon = packageInfo.applicationInfo.loadIcon(packageManager);
-                holder.imageAppIcon.setImageDrawable(appIcon);
+                if (packageInfo != null) {
```

Comment:
Should this ever be `null`?

Reply:
The Ion library creates an APK file right away when starting a new download.
During the download is progressing, if app list is scrolled up and down, onBindViewHolder will be called, which triggers app icon extraction to display in app list. That may cause the package info missing (meaning null) because the APK file is not fully downloaded unless download's finished.